text/0000-stable-simd.md

```diff
@@ -0,0 +1,720 @@
+write `extern "platform-intrinsics" { ... }` or `#[link_llvm_intrinsic...]`.
+This is certainly a *much* smaller surface area to stabilize (aka not
+thousands of intrinsics).
+
+This avenue was decided against, however, for a few reasons:
+
+* Such raw interfaces may change over time as they simply represent LLVM as a
+  current point in time rather than what LLVM wants to do in the future.
+* Alternate implementations of rustc or alternate rustc backends like Cretonne
+  may not expose the same sort of functionality that LLVM provides, or
+  implementing the interfaces may be much more difficult in alternate backends
+  than in LLVM's.
+
+As a result, it's intended that instead of exposing raw building blocks (and
+allowing `stdsimd` to live on crates.io) we'll instead pull in `stdsimd` to the
+standard library and expose it as the stable interface to SIMD in Rust.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+There's a number of unresolved questions around stabilizing SIMD today which
+don't pose serious blockers and may also wish to be considered open bugs rather
+than blocking stabilization:
+
+## Relying on unexported LLVM APIs
+
+The static detection performed by `cfg!` and `#[cfg]` currently relies on a
+[Rust-specific patch to LLVM][llvm-patch]. LLVM internal knows all about
+hierarchies of features and such. For example if you compile with `-C
+target-feature=+avx2` then `cfg!(target_feature = "sse2")` also needs to resolve
+to `true`. Rustc, however, does not know about these features and relies on
+learning this information through LLVM.
+
+Unfortunately though LLVM does not actually export this information for us to
+consume (as far as we know). As a result we have a [local patch][llvm-patch]
+which exposed this information for us to read. The consequence of this
+implementation detail is that when compiled against the system LLVM the `cfg!`
+macro may not work correctly when used in conjunction with `-C target-feature`
+or `-C target-cpu` flags.
+
+It appears that clang [vendors and/or duplicates][clang] LLVM's functionality in
+this regard. It's an option for rustc to do the same but it may also be an
+option to expose the information in upstream LLVM. So far there appears to have
+been no attempts to upstream this patch into LLVM itself.
+
+[llvm-patch]: https://github.com/rust-lang/llvm/commit/68e1e29618b2bd094d82faac16cf8e89959bbd68
+[clang]: https://github.com/llvm-mirror/clang/blob/679d846fcc73bd213347785185006d591698a132/lib/Basic/Targets/X86.cpp
+
+## Intrinsics that take constant arguments and function pointers
+
+Mentioned above it's planned to give intrinsics that take constant arguments
+some special treatment, but unfortunately we still haven't forbidden function
+pointers to these intrinsics. That means that it's still possible to pass
+runtime-determined values which may otherwise not be possible to do in a world
+where it's powered by `const` features rather than special attributes.
```

Comment:
Could we temporarily prohibit function pointer creation for functions tagged with `rustc_args_require_const`? Seems reasonably easy to implement and would ensure forward compatibility in this respect.

Reply:
There's a problem with not only `fn` pointers but also `Fn`* impls.

Reply:
Hm, bummer. Could we (hackily) make function types with that attribute not implement the Fn* traits? Might cause the same kind of trouble as we've long had with primitive `Copy` types not having `Clone` impls. But at least ICEs would be better than committing to a wrong interface.

Reply:
It should work without ICEs - the problem with `Copy`/`Clone` was that they'd implement `Copy` built-in but `Copy: Clone` so the compiler would assume `Clone` was available but then couldn't find an actual `impl`.

I don't know if there are problems with not implementing `FnOnce` on something callable. cc @nikomatsakis

Reply:
I suspect there'd be some ICE somewhere along the line...but probably could be done.

Reply:
I forget how much we've built on the assumption that all callable things implement the trait.

Reply:
Oh, another issue, the type of the `fn` item doesn't include the constant value now, but would in the future.

Which means we shouldn't allow using the function name outside of a call expression - that would alleviate all of the pointer/trait concerns, since the interactions would simply not exist!

Reply:
Oh yeah, a syntactic restriction seems like the best solution.

Reply:
@eddyb would you be able to help me implement that in the compiler? @rkruppe is right in that we basically just want to treat any function tagged `#[rustc_args_require_const]` this way, I just wasn't sure where to do this in rustc!

Reply:
For `foo(...)`, you'd want to check in `rustc_typeck::check::FnCtxt::instantiate_value_path`, at some point, that, if `def.def_id()` has `#[rustc_args_require_const]`, `tcx.hir.get(tcx.hir.get_parent_node(node_id))` is an `NodeExpr` containing a `ExprCall(ref callee, ..)` *and* that `node_id` matches `callee.id` (i.e. it's not an argument in the call, but the function being called).

Methods might be trickier - do you want to support the attribute being applied to a trait/inherent method? I'd suggest against it, for simplicity's sake (so you'd want to stick some code somewhere, maybe in `ast_validation`) that only allows the attribute on non-associated `fn`s.

Reply:
Thanks @eddyb! Now that I've got https://github.com/rust-lang/rust/pull/48078 working I'll move this to the design rather than an unresolved question.